### DISCUSSION TITLE
python dbus library comparison

### DISCUSSION THREAD

**Discussion by fohrloop** [fohrloop/wakepy#47] [General]
Just wanted to write down some of my notes from comparison of multiple different python dbus implementations. The meaning of the column names are:
- **DL / day**: Downloads per day through PyPI. Very rough and non-scientific estimate, done by eyeballing the pypistats.org graph, near maximum of weekday downloads on 2023-06-17.
- **Installs**: Tested on fresh Ubuntu 22.04 install, on fresh python virtual environment, with *simple pip install \<package\>*.
- **Works**: If the dbus python package actually works in the fresh virtual environment. Some packages will install without issues but will throw an exception during runtime (e.g. for a missing dependency, like PyGObject)
- **Works w/o ADDR**: Could use session bus even if `DBUS_SESSION_BUS_ADDRESS` environment variable would be missing. Not sure if this is good thing or not (where is the address then taken from..?)
- **Tech**: Some short words about the underlying technology
- **Licence**: The Licence for the package.

|                      |    DL / day    | Installs | Works | Work w/o ADDR | Tech                   | Licence? |
|----------------------|---------|----------|-------|---------------|------------------------|----------|
| [jeepney](https://gitlab.com/takluyver/jeepney)   ([docs](https://jeepney.readthedocs.io/))           | 850000 | Yes      | Yes   | No            | socket[1]              | MIT      |
| [dbus-python](https://salsa.debian.org/debian/dbus-python) ([docs](https://dbus.freedesktop.org/doc/dbus-python/index.html))          | 5400   | No       |    N/A   |    N/A            | dbus, libdbus[2]       | multiple |
| [dbus-next](https://github.com/altdesktop/python-dbus-next)   ([docs](https://python-dbus-next.readthedocs.io/))        | 1250   | Yes      | Yes   | No            | socket+asyncio[3]      |   MIT       |
| [pydbus](https://github.com/LEW21/pydbus)     ([docs](https://pydbus.readthedocs.io/))[5]   | 800     | Yes      | No    | N/A            | PyGObject [7]              |   LGPL-2.1       |
| [dasbus](https://github.com/rhinstaller/dasbus) ([docs](https://dasbus.readthedocs.io/))               | 100     | Yes      | No    | N/A            | PyGObject [7]             |   LGPL-2.1       |
| [sdbus](https://github.com/python-sdbus/python-sdbus) ([docs](https://python-sdbus.readthedocs.io/))[6] | 100     | Yes      | Yes   | Yes           | libsystemd/libelogind[4]          |  LGPL-2.1-or-later       |
| [dbussy](https://github.com/ldo/dbussy)               | 20      | Yes      | Yes   | Yes           | ctypes+libdbus-1.so.3  |    LGPL-2.1      |
| [pysdbus](https://github.com/anyc/pysdbus)              | -       |   (Yes)       | Yes   | Yes           | ctypes+libsystemd.so.0 |     LGPL-2.1     |
| [pystemd](https://github.com/systemd/pystemd)              |         | No       |    N/A   |       N/A      |      cython wrapper for libsystemd               |        LGPL-2.1  |

#### Footnotes
[1] Zero dependency (python or any other) package
[2] Required libdbus. Has .c and .h files in source, so needs to be compiled.
[3] Zero dependency (python or any other) package on `dbus_next.aio` (async side)
[4] sdbus wheel in PyPi is statically linked with libsystemd, but it is also possible to compile against libelogind (for Alpine/other elogind based systems). See [this](https://github.com/fohrloop/wakepy/discussions/47#discussioncomment-8950282) comment. Does not use ctypes or dynamic linking.
[5] Abandoned? Last commit from 2018.
[6] Aka. python-sdbus
[7] Note that PyGObject (aka. PyGI or python-gi. Imported as `gi`) does not install easily, either. The dependencies are listed in the [documentation](https://pygobject.readthedocs.io/en/latest/getting_started.html). 

#### Yet other alternatives
- There is also txdbus, which is really only intended to be used with twisted. It even automatically installs twisted, when installing txdbus.
- There would also be possibility to use PyGObject (PyGI / gi) directly, or as said in [dbus-python docs](https://dbus.freedesktop.org/doc/dbus-python/), using GDBus, part of the GIO module of [GLib](http://developer.gnome.org/glib/), or QtDBus, part of Qt, via [PyQt](http://www.riverbankcomputing.co.uk/software/pyqt/intro).

### Comments

- The only python only dbus packages with zero python (or other) dependencies are jeepney and dbus-next, which both use the python [socket](https://docs.python.org/3/library/socket.html) module from the standard library. 
- The only packages with MIT licences are also jeepney and dbus-next.
- Jeepney has much larger install base, and in fact, it was also installed with Ubuntu 22.04 (As were also dbus-python and PyGObject / gi)
- Installs: Note that this column means, "installs on fresh Ubuntu 22.04 install, inside fresh virtual environment, with a single `pip install` command". In other words, installs well without a need for installing e.g. non-python dependencies. Most packages with other than python (e.g C) code, if not delivering wheels for the system, will have difficulties when installing. A package which would *not* install without effort in new virtual environment is not very appealing candidate for a dependency (or dependency of a dependency..)

<details>
<summary>Edits to this post</summary>
2024-04-29: Corrected the python-sdbus license to LGPL-2.1-or-later. Added note about libelogind option.
</details>

**Comment by fohrloop** [comment]
Here are some (raw) test results from at least some of the packages tested. For the tests, I used the following D-Bus service, object, interface and method:

- dest / service / bus_name: org.freedesktop.DBus
- object path: /org/freedesktop/DBus
- interface: org.freedesktop.DBus
- method: ListNames

Test for need for `DBUS_SESSION_BUS_ADDRESS` was performed by adding

```python
import os
del os.environ['DBUS_SESSION_BUS_ADDRESS']
```
to the top of the test script.


## Results

<details>
  <summary>jeepney</summary>

### Jeepney

```python
from jeepney import DBusAddress, new_method_call
from jeepney.io.blocking import open_dbus_connection

proxy = DBusAddress('/org/freedesktop/DBus',
                            bus_name='org.freedesktop.DBus',
                            interface='org.freedesktop.DBus')

connection = open_dbus_connection(bus='SESSION')

msg = new_method_call(proxy, 'ListNames')

reply = connection.send_and_get_reply(msg)
print(reply.body[0])
```

- Installs well
- Works
- Does not work if DBUS_SESSION_BUS_ADDRESS is unset

```
>>> connection = open_dbus_connection(bus='SESSION')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/jeepney/io/blocking.py", line 341, in open_dbus_connection
    bus_addr = get_bus(bus)
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/jeepney/bus.py", line 53, in get_bus
    return find_session_bus()
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/jeepney/bus.py", line 42, in find_session_bus
    addr = os.environ['DBUS_SESSION_BUS_ADDRESS']
  File "/usr/lib/python3.10/os.py", line 679, in __getitem__
    raise KeyError(key) from None
KeyError: 'DBUS_SESSION_BUS_ADDRESS'
```
</details>

<details>
  <summary>dbus-python</summary>

### dbus-python


Installing in venv does not work directly:

```
(venv) np8@np8-HP-EliteBook-840-G5:~/tmp$ python -m pip install dbus-python
Collecting dbus-python
  Downloading dbus-python-1.3.2.tar.gz (605 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 605.5/605.5 KB 11.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      + meson setup /tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9 /tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9/.mesonpy-081b9hy0/build -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9/.mesonpy-081b9hy0/build/meson-python-native-file.ini
      The Meson build system
      Version: 1.1.1
      Source dir: /tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9
      Build dir: /tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9/.mesonpy-081b9hy0/build
      Build type: native build
      Project name: dbus-python
      Project version: 1.3.2
      
      ../../meson.build:4:0: ERROR: Unknown compiler(s): [['cc'], ['gcc'], ['clang'], ['nvc'], ['pgcc'], ['icc'], ['icx']]
      The following exception(s) were encountered:
      Running `cc --version` gave "[Errno 2] No such file or directory: 'cc'"
      Running `gcc --version` gave "[Errno 2] No such file or directory: 'gcc'"
      Running `clang --version` gave "[Errno 2] No such file or directory: 'clang'"
      Running `nvc --version` gave "[Errno 2] No such file or directory: 'nvc'"
      Running `pgcc --version` gave "[Errno 2] No such file or directory: 'pgcc'"
      Running `icc --version` gave "[Errno 2] No such file or directory: 'icc'"
      Running `icx --version` gave "[Errno 2] No such file or directory: 'icx'"
      
      A full log can be found at /tmp/pip-install-xk_10dzx/dbus-python_dd5c78b7444e4336845eded9dc9564d9/.mesonpy-081b9hy0/build/meson-logs/meson-log.txt
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

</details>

<details>
  <summary>PyGObject</summary>

### PyGObject

- Has some dependencies listed in the [documentation](https://pygobject.readthedocs.io/en/latest/getting_started.html).
- Without installing dependencies, the install from pip will fail (as expected):
  
```
(venv) np8@np8-HP-EliteBook-840-G5:~/tmp$ python -m pip install PyGObject
Collecting PyGObject
  Downloading PyGObject-3.44.1.tar.gz (720 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 720.5/720.5 KB 8.8 MB/s eta 0:00:00
  Installing build dependencies ... error
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [41 lines of output]
      Collecting setuptools
        Using cached setuptools-68.0.0-py3-none-any.whl (804 kB)
      Collecting wheel
        Using cached wheel-0.40.0-py3-none-any.whl (64 kB)
      Collecting pycairo
        Downloading pycairo-1.24.0.tar.gz (344 kB)
           ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 344.9/344.9 KB 5.5 MB/s eta 0:00:00
        Installing build dependencies: started
        Installing build dependencies: finished with status 'done'
        Getting requirements to build wheel: started
        Getting requirements to build wheel: finished with status 'done'
        Installing backend dependencies: started
        Installing backend dependencies: finished with status 'done'
        Preparing metadata (pyproject.toml): started
        Preparing metadata (pyproject.toml): finished with status 'done'
      Building wheels for collected packages: pycairo
        Building wheel for pycairo (pyproject.toml): started
        Building wheel for pycairo (pyproject.toml): finished with status 'error'
        error: subprocess-exited-with-error
      
        × Building wheel for pycairo (pyproject.toml) did not run successfully.
        │ exit code: 1
        ╰─> [12 lines of output]
            running bdist_wheel
            running build
            running build_py
            creating build
            creating build/lib.linux-x86_64-cpython-310
            creating build/lib.linux-x86_64-cpython-310/cairo
            copying cairo/__init__.py -> build/lib.linux-x86_64-cpython-310/cairo
            copying cairo/__init__.pyi -> build/lib.linux-x86_64-cpython-310/cairo
            copying cairo/py.typed -> build/lib.linux-x86_64-cpython-310/cairo
            running build_ext
            'pkg-config' not found.
            Command ['pkg-config', '--print-errors', '--exists', 'cairo >= 1.15.10']
            [end of output]
      
        note: This error originates from a subprocess, and is likely not a problem with pip.
        ERROR: Failed building wheel for pycairo
      Failed to build pycairo
      ERROR: Could not build wheels for pycairo, which is required to install pyproject.toml-based projects
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```
</details>

<details>
  <summary>dbus-next</summary>

### dbus-next

```python
import asyncio
from dbus_next.aio import MessageBus
from dbus_next.message import Message

async def main():
    bus = await MessageBus().connect()
    msg = Message(destination='org.freedesktop.DBus',
        path='/org/freedesktop/DBus',
        interface='org.freedesktop.DBus',
        member='ListNames')
    return await bus.call(msg)
    

reply = asyncio.run(main())
print(reply.body[0])
```

- Works when DBUS_SESSION_BUS_ADDRESS is set
- Does not work without it:

```
... 
>>> reply = asyncio.run(main())
Traceback (most recent call last):
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/dbus_next/_private/address.py", line 78, in get_session_bus_address
    with open(dbus_info_file_name) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/np8/.dbus/session-bus/e76b5954394a44a2a46b73c3b2ff31ba-0'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "<stdin>", line 2, in main
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/dbus_next/aio/message_bus.py", line 122, in __init__
    super().__init__(bus_address, bus_type, ProxyObject)
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/dbus_next/message_bus.py", line 72, in __init__
    get_bus_address(bus_type))
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/dbus_next/_private/address.py", line 98, in get_bus_address
    return get_session_bus_address()
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/dbus_next/_private/address.py", line 81, in get_session_bus_address
    raise InvalidAddressError(f'could not open dbus info file: {dbus_info_file_name}')
dbus_next.errors.InvalidAddressError: could not open dbus info file: /home/np8/.dbus/session-bus/e76b5954394a44a2a46b73c3b2ff31ba-0
```


- Note that the "glib" MessageBus requires gi (PyGObject):

```python
In [1]: from dbus_next.glib import MessageBus

In [2]: MessageBus()
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[2], line 1
----> 1 MessageBus()

File ~/tmp/venv/lib/python3.10/site-packages/dbus_next/glib/message_bus.py:154, in MessageBus.__init__(self, bus_address, bus_type, auth)
    149 def __init__(self,
    150              bus_address: str = None,
    151              bus_type: BusType = BusType.SESSION,
    152              auth: Authenticator = None):
    153     if _import_error:
--> 154         raise _import_error
    156     super().__init__(bus_address, bus_type, ProxyObject)
    157     self._main_context = GLib.main_context_default()

File ~/tmp/venv/lib/python3.10/site-packages/dbus_next/glib/message_bus.py:17
     15 _import_error = None
     16 try:
---> 17     from gi.repository import GLib
     18     _GLibSource = GLib.Source
     19 except ImportError as e:

ModuleNotFoundError: No module named 'gi'
```
</details>

<details>
  <summary>pydbus</summary>

### pydbus

```python
from pydbus import SessionBus

session_bus = SessionBus()
        
proxy = bus.get('org.freedesktop.DBus',
                '/org/freedesktop/DBus')
interface = proxy['org.freedesktop.DBus']
interface.ListNames()
```

Depends on PyGObject:

```
>>> from pydbus import SessionBus
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/pydbus/__init__.py", line 1, in <module>
    from .bus import SystemBus, SessionBus, connect
  File "/home/np8/tmp/venv/lib/python3.10/site-packages/pydbus/bus.py", line 1, in <module>
    from gi.repository import Gio
ModuleNotFoundError: No module named 'gi'
```

- The requirements are listed for example in [this dockerfile](https://github.com/hcoin/pydbus/blob/master/tests/py3.5-ubuntu-16.04.dockerfile) as mentioned in the [docs](https://pydbus.readthedocs.io/en/latest/installation.html).

```
apt-get install -y dbus python3-gi python3-pip psmisc dbus-x11 python3-dev dbus-x11 python-gi-dev python3  libglib2.0 libglib2.0-dev gobject-introspection
pip3 install greenlet
```
</details>

<details>
  <summary>sdbus</summary>

### sdbus

  
```python
from sdbus import DbusInterfaceCommon, dbus_method

class Interface(DbusInterfaceCommon, 
                interface_name='org.freedesktop.DBus'):

    @dbus_method()
    def ListNames(self) -> str:
        raise NotImplementedError           

iface = Interface(
    service_name='org.freedesktop.DBus',
    object_path='/org/freedesktop/DBus',
)


print(iface.ListNames())
```

- Installs from pip easily
- Works even without DBUS_SESSION_BUS_ADDRESS

</details>

<details>
  <summary>dasbus</summary>

### dasbus

- Requires PyGObject (also mentioned on the [README](https://github.com/rhinstaller/dasbus/tree/master))

```
>>> from dasbus.connection import SessionMessageBus
>>> bus = SessionMessageBus()
...
File ~/tmp/venv/lib/python3.10/site-packages/dasbus/typing.py:27
      1 #
      2 # Support for DBus types
      3 #
   (...)
     22 # https://dbus.freedesktop.org/doc/dbus-specification.html#type-system.
     23 #
     25 from typing import Tuple, Dict, List, NewType
---> 27 import gi
     28 gi.require_version("GLib", "2.0")
     29 from gi.repository.GLib import Variant, VariantType

ModuleNotFoundError: No module named 'gi'
```
</details>

<details>
  <summary>dbussy</summary>

### dbussy

```python
import ravel # part of dbussy

conn = ravel.session_bus()
peer = conn['org.freedesktop.DBus']
obj = peer['/org/freedesktop/DBus']
iface = obj.get_interface('org.freedesktop.DBus')
results = iface.ListNames()
print(results)
```

- Works also without DBUS_SESSION_BUS_ADDRESS !

</details>

<details>
  <summary>pystemd</summary>

### pystemd

- Installing from pip gives an error:

```
(venv) np8@np8-HP-EliteBook-840-G5:~/tmp$ python -m pip install pystemd
Collecting pystemd
  Downloading pystemd-0.13.2.tar.gz (308 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 308.4/308.4 KB 6.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      "pkg-config" command could not be found. Please ensure "pkg-config" is installed into your PATH.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

</details>

<details>
  <summary>pysdbus</summary>

### pysdbus

- Ctypes wrapper for libsystemd.so.0
  
```python
import dbus

bus = dbus.SessionBus()
obj = bus.getObject("org.freedesktop.DBus", "/org/freedesktop/DBus")
iface = obj.getInterface("org.freedesktop.DBus")
reply = iface.ListNames()
print(reply)
```

- Works, also without DBUS_SESSION_BUS_ADDRESS !

</details>

**Comment by igo95862** [comment]
Hello, author of python-sdbus here.

> libsystemd[4]

Python-sdbus also supports the elogind when available during compile time. (meaning Alpine Linux is supported) The PyPI package however is linked against glibc so without compiling during installation only glibc based distros are available.

> LGPL-2.1, GPL-2.0

The license is LGPL-2.1-or-later but the GPL-2.0 text is included because LGPL is an extension of GPL and requires the text of GPL to be present.

**Reply by fohrloop** [reply]
Hi @igo95862, thank you so much for the corrections. I updated the list, and hope it is more accurate now.
